concurrency_test.go

```diff
@@ -0,0 +1,270 @@
+package statsig
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestCallingAPIsConcurrently(t *testing.T) {
+	flushedEventCount := 0
+	testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		res.WriteHeader(http.StatusOK)
+		if strings.Contains(req.URL.Path, "download_config_specs") {
+			var in *downloadConfigsInput
+			bytes, _ := ioutil.ReadFile("download_config_specs.json")
+			json.NewDecoder(req.Body).Decode(&in)
+			res.Write(bytes)
+		} else if strings.Contains(req.URL.Path, "log_event") {
+			type requestInput struct {
+				Events          []Event         `json:"events"`
+				StatsigMetadata statsigMetadata `json:"statsigMetadata"`
+			}
+			input := &requestInput{}
+			defer req.Body.Close()
+			buf := new(bytes.Buffer)
+			buf.ReadFrom(req.Body)
+
+			json.Unmarshal(buf.Bytes(), &input)
+			flushedEventCount += len(input.Events)
+		} else if strings.Contains(req.URL.Path, "get_id_lists") {
+			baseURL := "http://" + req.Host
+			r := map[string]idList{
+				"list_1": {Name: "list_1", Size: 3, URL: baseURL + "/list_1", CreationTime: 1, FileID: "file_id_1"},
+				"list_2": {Name: "list_2", Size: 3, URL: baseURL + "/list_2", CreationTime: 1, FileID: "file_id_2"},
+			}
+			v, _ := json.Marshal(r)
+			res.Write(v)
+		} else if strings.Contains(req.URL.Path, "list_1") {
+			res.Write([]byte("+7/rrkvF6\n"))
+		}
+	}))
+
+	defer testServer.Close()
+	options := &Options{
+		API: testServer.URL,
+		Environment: Environment{
+			Params: map[string]string{
+				"foo": "bar",
+			},
+			Tier: "awesome_land",
+		},
+	}
+
+	InitializeWithOptions("secret-key", options)
+
+	const (
+		goroutines = 10
+		loops      = 10
+	)
+
+	var wg sync.WaitGroup
+	wg.Add(goroutines)
+	for g := 0; g < goroutines; g++ {
+		go func() {
+			defer wg.Done()
+			statsigUser := User{
+				UserID:            fmt.Sprintf("statsig_u_%d", g),
```

Comment:
This is a data race here. The "g" variable is captured in a closure. Multiple goroutines are using it.

You could handle it in two ways. You could pass g as a parameter explicitly. Or you could do a self assignment to shadow the g variable. The latter is my preference. Looks like this.

```
for g := 0; g < goroutines; g++ {
     g := g
     go func() {
           // use the copy of g
     }()
}